File: algorithm_java/src/com/note/algorithm/sort/HeapSort.java
package com.note.algorithm.sort;

import com.note.algorithm.tool.Tool;

/**
 * 堆排序
 */
public class HeapSort
{

    public static void main(String[] args)
    {

    }

    /**
     * 向上调整
     * @param arr 原始数组
     * @param index 新插入的元素的下标索引
     */
    static void heapInsert(int[] arr, int index)
    {
        while (arr[index] > arr[(index - 1) / 2])
        {
            int temp = arr[index];
            arr[index] = arr[(index - 1) / 2];
            arr[(index - 1) / 2] = temp;
            index = (index - 1) / 2;
        }
    }

    /**
     * 向下调整
     * @param arr 数组
     * @param index 变化元素的索引
     * @param size 数字的尺寸
     */
    static void heapAdjust(int[] arr, int index, int size)
    {
        int l = index * 2 + 1;
        while (l < size)
        {
            int best = l + 1 < size && arr[l] > arr[l + 1]
                    ? l + 1 : l;
            best = arr[best] > arr[index] ? best : index;
            if (best == index) break;
            Tool.swap(arr, index, best);
            index = best;
            l = index * 2 + 1;
        }
    }

    /**
     * 从顶到底建堆
     * @param arr 原始数组
     */
    public static void heapSort1(int[] arr)
    {
        int n = arr.length;
        //建堆
        for (int i = 0; i < n; i++)
        {
            heapInsert(arr, i);
        }
        int size = n;
        while (size > 1)
        {
            Tool.swap(arr, 0, --size);
            heapAdjust(arr, 0, size);
        }
    }

    /**
     * 从底到顶建堆
     * @param arr 原始数组
     */
    public static void heapSort2(int[] arr)
    {
        int n = arr.length;
        for (int i = n - 1; i >= 0; i--)
        {
            heapAdjust(arr, i, n);
        }
        int size = n;
        while (size > 1)
        {
            Tool.swap(arr, 0, --size);
            heapAdjust(arr, 0, size);
        }
    }

}

File: algorithm_cpp/sort/heap_sort.cpp
//
// Created by qi wang on 2025/7/13.
//

#include <iostream>
void heap_insert(int arr[], int index);
void swap(int arr[], int i, int j);
void heap_adjust(int arr[], int index, int size);
void heap_sort(int arr[], int size);

int main()
{
    int arr[] = { 3, 11, 53, 23, 11, 21, 55, 43 };
    heap_sort(arr, 8);
    for (int i = 0; i < 8; i++)
        std::cout << arr[i] << " ";
    std::cout << std::endl;
}

/**
 * 交换数组的两个元素
 * @param arr
 * @param i
 * @param j
 */
void swap(int arr[], int i, int j)
{
    int temp = arr[i];
    arr[i] = arr[j];
    arr[j] = temp;
}

/**
 * 向上调整
 * @param arr 原始数组
 * @param index 最后一个元素的索引
 */
void heap_insert(int arr[], int index)
{
    while (arr[index] > arr[(index - 1) / 2])
    {
        swap(arr, index, (index - 1) / 2);
        index = (index - 1) / 2;
    }
}

/**
 * 向下调整
 * @param arr
 * @param index
 * @param size
 */
void heap_adjust(int arr[], int index, int size)
{
    int l = index * 2 + 1;
    while (l < size)
    {
        int best = l + 1 < size  && arr[l] < arr[l + 1]
            ? l + 1
            : l;
        best = arr[index] > arr[best] ? index : best;
        if (best == index) break;
        swap(arr, index, best);
        index = best;
        l = index * 2 + 1;
    }
}

/**
 * 从顶到底建堆
 * @param arr
 * @param size
 */
void heap_sort(int arr[], int size)
{
    //建堆
    for (int i = 0; i < size; i++)
    {
        heap_insert(arr, i);
    }
    //排序
    int n = size;
    while (n > 1)
    {
        swap(arr, 0, --n);
        heap_adjust(arr, 0, n);
    }
}
